### PR TITLE
ports - sav

### PR DESCRIPTION
**Pull request by qqdipps**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. |  class methods for top_ten and spotlight in work, also private helper method for those sort work by votes. 
Describe how you approached testing that model method. What edge cases did you come up with? | I kept with the validation, relationship, custom method breakdown for each model.  Edge case for top_ten includes if no works in db and less than 10 works in db.
What are `session` and `flash`? What is the difference between them? | Session and flash are both hash like structures that are stored in rails and involve cookie. The main difference is that flash persists for a single response/request cycle while session persists for many. 
What was one thing that you gained more clarity on through this assignment? | I gained more clarity on complex relationships in db, thinking about design and how it can be extended. I feel like I did lots of cut and paste in my views and could dry them up using view helpers. Also explored different work flow.
What is the Heroku URL of your deployed application? | https://trello.com/b/8A6CUAJP/mediaranker https://green-media-ranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | can't think of any right now, but I'll update this if I think of anything =).

### CODE REVIEW

File: test/models/user_test.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    end
+
+    it "will not be valid with same username twice" do
+      user_new = User.new(username: user.username)
+      expect(user_new.save).must_equal false
+    end
+  end
+
+  describe "relations" do
+    it "will have 0 votes" do
```
Comment (CheezItMan) [inline test/models/user_test.rb:23]: probably should be `it "can have 0 votes" do`

File: app/models/vote.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+class Vote < ApplicationRecord
+  belongs_to :user
+  belongs_to :work
+
+  validates :user, presence: true, uniqueness: { scope: [:work] }
```
Comment (CheezItMan) [inline app/models/vote.rb:5]: Nice!

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        flash[:success] = "Successfully created new user #{@user.username} new with ID #{@user.id}"
+      else
+        flash.now[:error] = "A problem occurred: Could not log in"
+        @user.errors.messages.each do |label, message|
+          flash.now[label.to_sym] = message
+        end
+        render :login_form, status: :bad_request
+      end
+    end
+    if @user.valid?
```
Comment (CheezItMan) [inline app/controllers/users_controller.rb:34]: This if statement seems a little extra, maybe if you had an earlier `return` when the user isn't valid you wouldn't need this if.

File: app/controllers/votes_controller.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
+class VotesController < ApplicationController
+  def create
```
Comment (CheezItMan) [inline app/controllers/votes_controller.rb:2]: Take a look at how big this method is.  I suggest creating a method `upvote` on the work or user model.  That might help you simplify this method a bit.

File: app/controllers/works_controller.rb [hunk 5]
```diff
@@ -0,0 +1,70 @@
+class WorksController < ApplicationController
+  before_action :find_work, except: [:index, :new, :create]
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:2]: Nice a filter!

File: app/controllers/works_controller.rb [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    end
+  end
+
+  def destroy
+    if @work
+      flash[:success] = "Successfully destroyed #{@work.category} #{@work.id}"
+      @work.destroy
+      redirect_to root_path
+    else
+      head :not_found
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:57]: Maybe instead of `head` now we should do a `redirect` with an error message to the user with flash.

File: test/controllers/votes_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
+require "test_helper"
+
+describe VotesController do
+  let(:user) { users(:three) }
+  let(:work) { works(:one) }
+  describe "Votes#create" do
+    it "will create a vote if logged in" do
```
Comment (CheezItMan) [inline test/controllers/votes_controller_test.rb:7]: You need a test for the action if they're not logged in, and a test for a duplicate vote

File: test/controllers/works_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    end
+  end
+
+  describe "works#update" do
+    it "will update work with valid params" do
+      work_param = { work: { title: "new work",
+                           category: "book",
+                           id: work.id } }
+      expect {
+        post works_path, params: work_param
```
Comment (CheezItMan) [inline test/controllers/works_controller_test.rb:97]: Wait a minute!  this is a `post` request not a `patch`!!!!!!

File: test/controllers/works_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+      }.must_change "Work.count", 1
```
Comment (CheezItMan) [inline test/controllers/works_controller_test.rb:98]: Also `update` shouldn't be changing the count of works!

**Comment by CheezItMan** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | Good number of commits and good commit messages
Comprehension questions |  Check
**General** |
Rails fundamentals (RESTful routing, use of named paths) |  Check
Views are well-organized (DRY, use of semantic HTML, use of partials) |  Check
Errors are reported to the user |  Check, although double voting does 
Business logic lives in the models |  Sure, for the most part, see my inline notes on adding an upvote to the models.  
Models are thoroughly tested, including relations, validations and any custom logic |  Pretty good model testing
Controllers are thoroughly tested, including the login/logout process and multi-step workflows like voting for a work |  You've got some problems with your `update` action in the works controller.
**Wave 1 - Media** |
Splash page shows the three media categories |  Check
Basic CRUD operations on media are present and functional |Check
**Wave 2 - Users and Votes** |
Users can log in and log out |Check
The ID of the current user is stored in the session |Check
A user cannot vote for the same media more than once |Check
All media lists are ordered by vote count |Check
Splash page contains a media spotlight |Check
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users |Check
Individual user pages and the user list are present |Check
**Optional - Styling** |
Bootstrap is used appropriately |Check
Look and feel is similar to the original |  Looks great!
**Overall** |  Nice work, you hit all the learning goals for the project.  You do need to check out my comments in the Works_controler_test.rb file.  You didn't actually test the `update` action.  Other than that, really great!
